website: plannable import docs (#33266)
hashicorp/terraform: website/docs/language/import/index.mdx

---
page_title: Import - Configuration Language
description: >-
  Import and manage existing resources with Terraform using configuration-driven import.
---

# Import

-> **Note:** Import blocks are only available in Terraform v1.5.0 and later.

> ~> **Experimental:** Import blocks are an experimental Terraform feature. While we do not expect to make backwards-incompatible changes to syntax, the processing of imports during the plan stage, generating configuration, and the `-generate-config-out` flag may change in future releases.

Reviewer: > Import blocks are an experimental Terraform feature

Did you mean to state this? or did you mean to only state that config gen is experimental?

The `import` block records that Terraform imported the resource and that Terraform did not previously create that resource with its configuration. Put another way, the `import` block records the user's intent to add an import to Terraform's plan graph.

## Syntax

The `import` block can appear in any Terraform configuration file. A common pattern is to create an `imports.tf` file, or to place each `import` block beside the `resource` block it imports into.

```hcl
import {
  to = aws_instance.example
  id = "i-abcd1234"
}

resource "aws_instance" "example" {
  name = "hashi"
  # (other resource arguments...)
}
```

The above `import` block defines an import of the AWS instance with the ID "i-abcd1234" into the `aws_instance.example` resource in the root module.

The `import` block has the following arguments:
 - `to` - The resource instance address this import is targeting.
 - `id` - A string with the import ID of the resource.

>  - `provider` (optional) - An optional custom resource provider, see [The Resource provider Meta-Argument](/terraform/language/meta-arguments/resource-provider) for details.

Reviewer: Should we state that without this, import{} will default to looking up the resource via the default provider?

---
page_title: Import - Configuration Language
description: >-
  Import and manage existing resources with Terraform using configuration-driven import.
---

# Import

-> **Note:** Import blocks are only available in Terraform v1.5.0 and later.

~> **Experimental:** While we do not expect to make backwards-incompatible changes to syntax, the processing of imports during the plan stage, generating configuration, and the `-generate-config-out` flag may change in future releases.

Use the `import` block to import existing infrastructure resources into Terraform, bringing them under Terraform's management. Configuration-driven import using `import` blocks is predictable, works with CICD pipelines, and allows you to preview an import operation before modifying state.

> Once imported, an existing resource is present in Terraform's configuration and state, much like any other resource Terraform created. You can then manage that resource like any other, updating its attributes and destroying it as part of a standard resource lifecycle.

Reviewer 2: ```suggestion
Once imported, Terraform tracks the resource in your state file. You can then manage the imported resource like any other, updating its attributes and destroying it as part of a standard resource lifecycle.
```

> Once imported, an existing resource is present in Terraform's configuration and state, much like any other resource Terraform created. You can then manage that resource like any other, updating its attributes and destroying it as part of a standard resource lifecycle.

Author: No, once import is complete the resource must be in configuration one way or another.
Either you used config generation, or you manually added a `resource` block alongside the `import` block.  

_Detail: Applying an import, like other planned actions (e.g. create), reconciles config and state. The CLI command `terraform import`, on the other hand, modifies state directly, allowing it to drift from config - it is more like the `state rm` and `state mv` state manipulation commands than a `plan` in that sense._
cc @rkoron007 FYI

If you try to import a resource without config and don't ask Terraform to generate it with the `-generate-config-out` flag, `terraform plan` will error:

```
$ terraform plan

Planning failed. Terraform encountered an error while generating this plan.

╷
│ Error: Import block target does not exist
│ 
│   on main.tf line 1:
│    1: import {
│ 
│ The target for the given import block does not exist. If you wish to automatically generate config for this resource, use the
│ -generate-config-out option within terraform plan. Otherwise, make sure the target resource exists within your configuration. For example:
│ 
│   terraform plan -generate-config-out=generated.tf
╵
```

> The `import` block records that Terraform imported a resource and that Terraform did not previously create that resource with its configuration. Put another way, the `import` block records the user's intent to add an import to Terraform's plan graph.

Author: It's up to the user, like the `moved` block. Either of these blocks can be safely deleted after apply.

The `moved` block has an additional use case - authors of public and shared modules can keep it in their config to help users migrate to newer module versions.

The circumstances and use cases for retaining `import` blocks are quite individual and might evolve over time (and combine with other parts of the Refactoring workflow). Is it sufficient for now to say that keeping the `import` block in code is optional?

Agree the second sentence is unclear, and I think unnecessary.

## Syntax

The `import` block can appear in any Terraform configuration file. A common pattern is to create an `imports.tf` file, or to place each `import` block beside the `resource` block it imports into.

```hcl
import {
  to = aws_instance.example
  id = "i-abcd1234"
}

resource "aws_instance" "example" {
  name = "hashi"
  # (other resource arguments...)
}
```

The above `import` block defines an import of the AWS instance with the ID "i-abcd1234" into the `aws_instance.example` resource in the root module.

The `import` block has the following arguments:
 - `to` - The resource instance address this import is targeting.
 - `id` - A string with the import ID of the resource.
 - `provider` (optional) - An optional custom resource provider, see [The Resource provider Meta-Argument](/terraform/language/meta-arguments/resource-provider) for details.

> If you do not give the `import` block a `provider` argument, Terraform attempts to import from the default provider by default. 

Reviewer 3: Adding on to Rita's suggestion:
```suggestion
If you do not set the `provider` argument, Terraform attempts to import from the default provider. 
```

> ## Plan and Apply an Import

Reviewer 2: ```suggestion
## Plan and apply an import
```

> Terraform processes the `import` block during the plan stage. Along with create, update, remove, and move, import is an action Terraform performs on a resource. Once a plan is approved, Terraform imports the resource into its state during the subsequent apply stage.

Author: A misplaced urge to educate the reader on Terraform's internals. Removed.

## Syntax

The `import` block can appear in any Terraform configuration file. A common pattern is to create an `imports.tf` file, or to place each `import` block beside the `resource` block it imports into.

```hcl
import {
  to = aws_instance.example
  id = "i-abcd1234"
}

resource "aws_instance" "example" {
  name = "hashi"
  # (other resource arguments...)
}
```

The above `import` block defines an import of the AWS instance with the ID "i-abcd1234" into the `aws_instance.example` resource in the root module.

The `import` block has the following arguments:
 - `to` - The resource instance address this import is targeting.
 - `id` - A string with the import ID of the resource.
 - `provider` (optional) - An optional custom resource provider, see [The Resource provider Meta-Argument](/terraform/language/meta-arguments/resource-provider) for details.

If you do not give the `import` block a `provider` argument, Terraform attempts to import from the default provider by default. 

### Import ID

To locate the resource you want to import, Terraform requires an import ID in the `id` argument. This argument must currently be supplied as a literal string.

> The import block's `id` argument is the resource's import ID. The identifier you use for the `id` argument is resource-specific. You can find the required ID in the [provider documentation](https://registry.terraform.io/browse/providers) for the resource you wish to import.

Author: This paragraph also appears in the Generating Configuration page (in the "Add the `import` Block" section). Should it be removed from there?

## Plan and Apply an Import

Terraform processes the `import` block during the plan stage. Along with create, update, remove, and move, import is an action Terraform performs on a resource. Once a plan is approved, Terraform imports the resource into its state during the subsequent apply stage.

> You can use `import` blocks by doing the following:

Reviewer 2: ```suggestion
To import a resource using `import` blocks, you must:
```

1. Define an `import` block for the resource(s).

> 1. Add a `resource` block in your configuration.

Reviewer 2: ```suggestion
1. Add a corresponding `resource` block to your configuration.
```

Reviewer 2: should we add a note here about the option to generate the config

> 1. Run `terraform plan` to see how Terraform plans to import the resource(s).

Reviewer 2: ```suggestion
1. Run `terraform plan` to review how Terraform will import the resource(s).
```

> 1. Apply the configuration and have Terraform import resources and update your Terraform state.

Reviewer 2: ```suggestion
1. Apply the configuration to import the resources and update your Terraform state.
```

> **Hands-on:** Try the [State Import](/terraform/tutorials/state/state-import?utm_source=WEBSITE&utm_medium=WEB_IO&utm_offer=ARTICLE_PAGE&utm_content=DOCS) tutorial.

> The `import` block is [_idempotent_](https://en.wikipedia.org/wiki/Idempotence), meaning that applying an import action and running another plan will not generate another import action. 

Reviewer 2: not sure about our conventions for linking to other sources. @rkoron007? 

```suggestion
The `import` block is [_idempotent_](https://en.wikipedia.org/wiki/Idempotence), so the import operation will only run once for any import block.
```

Reviewer 3: I was the one who suggested this link. I think folks not familiar with this word may want further examples of how to use it. It's a perfect word to use here though. 

We don't say anything about linking to outside resources in any of our style guides; and I think Wikipedia is reliable enough to link to here.

Author: It's true that the import operation will only run once for any import block--as long as the resource remains in state. I'd be wary of making any more general statement than that. I think we want to say something that means "_ceteris paribus_ applying a config containing an import block will only apply the import once".

> Terraform only needs to import a resource into a single address once. Attempting to import a resource into the same address again is a (harmless) no-op. For this reason, you can remove `import` blocks after applying or safely leave them in your configuration as a record of the resource's origin for future module maintainers. For more information on maintaining configurations over time, see [Refactoring](/terraform/language/modules/develop/refactoring).

Reviewer 2: ```suggestion
Terraform only needs to import a given resource once. Attempting to import a resource into the same address again is a harmless no-op. You can remove `import` blocks after completing the import or safely leave them in your configuration as a record of the resource's origin for future module maintainers. For more information on maintaining configurations over time, see [Refactoring](/terraform/language/modules/develop/refactoring).
```

> ## Resource Configuration

Reviewer 2: ```suggestion
## Resource configuration
```

> When Terraform imports a resource, it needs you to define that resource in configuration before saving it to state. You can do this by either writing a `resource` block or by [generating configuration](/terraform/language/import/generating-configuration). 

Reviewer 2: i want to rephrase this a little so the ordering is clear.

is the following correct? To prevent destroying the resource, you must create a corresponding `resource` block for any `import` block before running an `apply` for the imported resource. You can create resource configuration manually or generate it using Terraform (link)

Author: "To prevent destroying the resource" is possibly misleading, because Terraform will not let you get that far. If you add an `import` block without corresponding `resource` config, it will just fail at plan with an error telling you that.

Reviewer 3: Okay, I tried distilling the order Rita suggested with you above point @kmoe. How does this feel?
```suggestion
Before importing, you must add configuration for every resource you want Terraform to import. Otherwise, Terraform throws an error during planning, insisting you add resource configuration before it can successfully import. You can create resource configuration manually or [generate it using Terraform](/terraform/language/import/generating-configuration). 
```

> We recommend writing a `resource` block if you know the majority of a provider's [resource arguments](/terraform/language/resources/syntax#resource-arguments). For example, if the imported resource is similar to an existing resource whose configuration you can copy and paste from.

Reviewer 2: ```suggestion
We recommend writing a `resource` block if you know what most of the [resource's arguments](/terraform/language/resources/syntax#resource-arguments) will be. For example, your configuration may already contain a similar resource whose configuration you can copy and modify.
```

We recommend [generating configuration](/terraform/language/import/generating-configuration) when importing multiple resources or a single complex resource that you do not already have the configuration for.

### Add a `resource` block

> Add a `resource` block for the resource you are importing, ensuring the address matches the import block's target(i.e., `to`). 

Reviewer 2: ```suggestion
Add a `resource` block for the resource to import. The resource address must match the import block's `to` argument. 
```

```hcl
import {
  to = aws_instance.example
  id = "i-abcd1234"
}

resource "aws_instance" "example" {
  name = "renderer"
}
```

> ### Generate Configuration

Reviewer 2: ```suggestion
### Generate configuration
```

> ---

Reviewer 3: What file structure are you intending here? Do you want to have two pages, like so:
```
Import
-> Overview
-> Generating Configuration
```

If yes, can you update the `website/data/language-nav-data.json` to the following:
```
{
      "title": "Import",
      "routes": [
          { "title": "Overview", "path": "import/index" },
          {
            "title": "Generating Configuration",
            "path": "import/generating-configuration"
          }
        ]
  },
```

Reviewer 3: (Resolving this in favor of a discussion on the `language-nav-data.json` file itself.)

page_title: Import - Configuration Language
description: >-
  Import resources into Terraform to bring them under Terraform's management by writing `import` blocks in code.
---

# Import

-> **Note:** Import blocks are only available in Terraform v1.5.0 and later.

~> **Experimental:** Import blocks are an experimental Terraform feature. While we do not expect to make backwards-incompatible changes to their syntax, processing of imports during the plan, generating configuration, and the `-generate-config-out` flag may change in future releases.

> Terraform can import existing infrastructure resources, which brings them under Terraform's management. Once you have imported a resource, it will be present in the Terraform configuration and state, just like a resource you had created via Terraform. You can then manage it like any other resource, updating its attributes and destroying it as part of a normal resource lifecycle.

Author: Thank you, this is much clearer. I'd like to avoid suggesting that the `import` block is doing the "importing" itself. Terraform plans and applies actions such as import, whereas code blocks like `import` only _declare_ what the user expects to have been done.

A practitioner, on the other hand, can be said to import resources _using_ Terraform. Perhaps:
```suggestion
Terraform can be used to import existing infrastructure resources, which brings them under Terraform's management. Once you have imported a resource, it will be present in the Terraform configuration and state, just like a resource you had created via Terraform. You can then manage it like any other resource, updating its attributes and destroying it as part of a normal resource lifecycle.
```

An `import` block records that Terraform imported the resource, that it was not originally created due to this Terraform configuration. Considered another way, it records the user's intent that an import be added to Terraform's plan graph.

## `import` block syntax

The `import` block can appear in any Terraform configuration file. A common pattern is to create an `imports.tf` file, or to place each `import` block next to the `resource` block it will import into.

```hcl
import {
  to = aws_instance.example
  id = "i-abcd1234"
}
```

> The above `import` block defines an import of the AWS instance with ID "i-abcd1234" into the resource address `aws_instance.example` in the root module.

Author: Terraform really pushes the limits of active voice docs with its declarative language and implicit, graph-based operation.

Declaring a `resource` in config, for example, is not telling Terraform to create that resource, it is telling Terraform that your desired state-of-the-world is one in which that resource exists just so. If during the plan stage Terraform determines that it must create the resource in order for this to be true, it plans a create (n.) and during the apply stage creates (v.) the resource.

### Long meditation on configuration-driven behaviour

Configuration-driven move stretched this metaphor a little, as the `moved` block on its own is more concerned with Terraform's internal state than the state-of-the-world. Declaring a `moved` block in config is telling Terraform that the desired state is one in which a resource that used to be at the `from` address is now at the `to` address. If that still needs to happen, Terraform plans a move and then moves the resource.
Does a `moved` block "move" a resource? In my opinion, no--especially since if the resource is already at the new address, Terraform's graph walk outputs a no-op action instead of move. Terraform moves resources, but you can [Use Configuration to Move Resources](https://developer.hashicorp.com/terraform/tutorials/configuration-language/move-config). 

The `import` block has some similarities to the move case, but involves both state and state-of-the-world. Use Configuration to Import Resources, by writing an `import` block in config, which declares your intent that the targeted resource be understood to have been created outside Terraform and therefore the subject of an import (n.). This stilted and indirect technical language is very elegantly translated into understandable prose in your import tutorial (just skimmed the PR).

[As you can see, there is some inconsistency over whether the past participle (`moved`) or noun-that-can-be-mistaken-for-a-verb (`import`) is more appropriate here, which I would like to resolve in future releases. "The `import` block" just proved too compelling to force the issue now.]

### Conclusion

In terms of user intent, you declare in config what you want to be the case, and at worst what you want to have happened _at some point up to and including the next apply_. Only the graph's inscrutable interior, not the practitioner, is a valid source of an intent to perform any given action in particular.

The `import` block has the following arguments:
 - `to` - The resource instance address targeted by this import.
 - `id` - The import ID of the resource. String.
 - `provider` (optional) - A custom resource provider. See [The Resource provider Meta-Argument](/terraform/language/meta-arguments/resource-provider).

## Planning and applying an import

> The `import` block, like all Terraform configuration blocks, is processed during the `terraform plan` operation, which outputs a list of changes Terraform will make if you proceed to apply. Along with create, update, remove, and move, import is an action Terraform can perform on a resource. The resource is only added to the state during `terraform apply`.

Author: Really nice, thank you

> ### 1. Add the `import` block to configuration

Reviewer 3: I feel like these steps belong in a tutorial walkthough rather than docs; I suggest we remove these steps in favor of a bulleted list explaining the steps. 

```suggestion
You can use `import` blocks by doing the following:
1. Define an `import` block for the resource(s).
1. Add a `resource` block in your configuration.
1. Run `terraform plan` to see how Terraform plans to import the resource(s).
1. Apply the configuration and have Terraform import resources and update your Terraform state.

> **Hands-on:** Try the [State Import](/terraform/tutorials/state/state-import?utm_source=WEBSITE&utm_medium=WEB_IO&utm_offer=ARTICLE_PAGE&utm_content=DOCS) tutorial.
```

Reviewer 3: Also, do we need to add a step about refreshing state in this list?

Author: No, the state is refreshed automatically during plan.

  # aws_iot_thing.bar will be imported
    resource "aws_iot_thing" "bar" {
        arn               = "arn:aws:iot:eu-west-1:1234567890:thing/foo"
        attributes        = {}
        default_client_id = "foo"
        id                = "foo"
        name              = "foo"
        version           = 1
    }

Plan: 1 to import, 0 to add, 0 to change, 0 to destroy.
```

### 4. Apply the import
Run `terraform apply`. You should see output like the following:

```
$ terraform apply

aws_iot_thing.bar: Preparing import... [id=foo]
aws_iot_thing.bar: Refreshing state... [id=foo]

Terraform will perform the following actions:

  # aws_iot_thing.bar will be imported
    resource "aws_iot_thing" "bar" {
        arn               = "arn:aws:iot:eu-west-1:1234567890:thing/foo"
        attributes        = {}
        default_client_id = "foo"
        id                = "foo"
        name              = "foo"
        version           = 1
    }

Plan: 1 to import, 0 to add, 0 to change, 0 to destroy.
aws_iot_thing.bar: Importing... [id=foo]
aws_iot_thing.bar: Import complete [id=foo]

Apply complete! Resources: 1 imported, 0 added, 0 changed, 0 destroyed.

```

> ## Resource configuration

Reviewer 3: ```suggestion
## Resource Configuration
```

> When Terraform imports a resource, it needs the resource to be defined in configuration before saving it to the state. This means that when using `import` blocks, you have two options.

Reviewer 3: ```suggestion
When Terraform imports a resource, it needs you to define that resource in configuration before saving it to state. You can do this by either writing a `resource` block or by generating configuration. 
```

> ### Option: Write a `resource` block

Reviewer 3: ```suggestion
### Write a `resource` block
```

> Write a `resource` block for the resource you are importing, whose address matches the import block target. For example:

Reviewer 3: Can you explain why someone would do one or the other? It's helpful to add guidance for what we _recommend_ folks do.

```hcl
import {
  to = aws_instance.example
  id = "i-abcd1234"
}

resource "aws_instance" "example" {
  name = "renderer"
}
```

> ### Option: Generate configuration

Reviewer 3: It seems like this information is duplicated in the `generation-configuration.mdx` file? To keep things DRY, we should only have this information in one place. This article is short, so I think it'd be fine to keep it all in one place, but would love to know what you think!

Reviewer 3: I'll wait to review the following section until we know where it's going to live.

Terraform can generate HCL for resources which do not already exist in configuration. The HCL produced by Terraform is intended to act as a template, and contains Terraform's best guess at the appropriate value for each resource argument. You may find that you arrive at an ideal configuration by starting from the HCL generated by Terraform, removing some attributes, adjusting the value of others, and rearranging `resource` blocks into files and modules as appropriate.

To instruct Terraform to generate configuration during the plan, supply a file path with `-generate-config-out`. The file should not already exist. For example:

```
terraform plan -generate-config-out=generated_resources.tf
```

Terraform will generate configuration for any resources targeted by an `import` block that do not already exist in configuration, and write the HCL to `generated_resources.tf`.

Please see the [Generating Configuration](/terraform/language/import/generating-configuration) page for more details on config generation.

## Examples

> ### Import into module

Author: I saw your note on not introducing code blocks with sentence fragments. Would I repeat "The following example..." for each one?

  name = "hashi"
  # (other resource arguments...)
}
```

The above `import` block defines an import of the AWS instance with the ID "i-abcd1234" into the `aws_instance.example` resource in the root module.

The `import` block has the following arguments:
 - `to` - The resource instance address this import is targeting.
 - `id` - A string with the import ID of the resource.
 - `provider` (optional) - An optional custom resource provider, see [The Resource provider Meta-Argument](/terraform/language/meta-arguments/resource-provider) for details.

## Plan and Apply an Import

Terraform processes the `import` block during the plan stage. Along with create, update, remove, and move, import is an action Terraform performs on a resource. Once a plan is approved, Terraform imports the resource into its state during the subsequent apply stage.

You can use `import` blocks by doing the following:
1. Define an `import` block for the resource(s).
1. Add a `resource` block in your configuration.
1. Run `terraform plan` to see how Terraform plans to import the resource(s).
1. Apply the configuration and have Terraform import resources and update your Terraform state.

> > **Hands-on:** Try the [State Import](/terraform/tutorials/state/state-import?utm_source=WEBSITE&utm_medium=WEB_IO&utm_offer=ARTICLE_PAGE&utm_content=DOCS) tutorial.

Author: @rkoron007 I've applied your suggestion here but now lost the comment. Should I remove the numbered headings below (up to "## Resource Configuration") in favour of the link to the tutorial?

Author: (I've removed it as I think it makes sense just to link to the better tutorial!)
